File: playwright_axe.js
// playwright_axe.js
const { chromium } = require('playwright');

const URL = 'http://localhost:3000';

async function runAxe(url) {
  const browser = await chromium.launch({
    headless: true,
    args: ['--no-sandbox'],
  });

  const page = await browser.newPage();
  await page.goto(url, { waitUntil: 'networkidle' });

  // Inject axe-core
  await page.addScriptTag({
    url: 'https://cdnjs.cloudflare.com/ajax/libs/axe-core/4.9.1/axe.min.js',
  });

  // Run axe (all rules; remove runOnly to run everything)
  const results = await page.evaluate(async () => {
    return await axe.run(document, {
      runOnly: {
        type: 'tag',
        values: ['wcag2a', 'wcag2aa'], // Level A + AA only; delete runOnly to run all rules
      },
    });
  });

  await browser.close();

  const violations = results.violations;

  if (violations.length > 0) {
    const totalNodes = violations.reduce((n, v) => n + v.nodes.length, 0);
    console.log('\nAccessibility violations (' + totalNodes + ' problem' + (totalNodes === 1 ? '' : 's') + ')\n');

    violations.forEach((v) => {
      console.log( v.id + ' — ' + v.help);
      v.nodes.forEach((node) => {
        const hrefMatch = node.html.match(/href=["']([^"']+)["']/);
        const where = hrefMatch ? hrefMatch[1] : node.html.replace(/\s+/g, ' ').slice(0, 60) + (node.html.length > 60 ? '...' : '');
        console.log('  • ' + where);
      });
      console.log('');
    });
    process.exit(1);
  } else {
    console.log('No accessibility violations found.');
    process.exit(0);
  }
}

runAxe(URL).catch((err) => {
  console.error(err);
  process.exit(1);
});

File: playwright_axe.py
# playwright_axe.py
import asyncio
from playwright.async_api import async_playwright

URL = "http://localhost:3000"

async def run_axe(url: str):
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True, args=["--no-sandbox"])
        page = await browser.new_page()

        await page.goto(url, wait_until="networkidle")

        # Inject axe-core
        await page.add_script_tag(
            await page.add_script_tag(url="https://cdnjs.cloudflare.com/ajax/libs/axe-core/4.9.1/axe.min.js")
        )

        # Run axe
        results = await page.evaluate("""
            async () => {
                return await axe.run(document, {
                    runOnly: {
                        type: 'tag',
                        values: ['wcag2a', 'wcag2aa']
                    }
                });
            }
        """)

        await browser.close()

        violations = results["violations"]

        if violations:
            print(f"Found {len(violations)} accessibility violations\n")
            for v in violations:
                print(f"Rule: {v['id']}")
                print(f"Description: {v['help']}")
                print(f"Impact: {v['impact']}")
                for node in v["nodes"]:
                    print(f"  Target: {node['target']}")
                    print(f"  HTML: {node['html']}")
                print("---")
            raise SystemExit(1)
        else:
            print("No accessibility violations found ✅")
            raise SystemExit(0)

asyncio.run(run_axe(URL))